File: Packages/WhoopKitExperimental/Sources/WhoopKit/Framing.swift
//
//  Framing.swift
//  WhoopKit
//
//  Swift port of the Python framing.py codec for the WHOOP 5.0 / MG
//  "puffin" protocol. Platform-pure (no CoreBluetooth import) so it's
//  testable without hardware, matching NOOP's own architecture.
//
//  Outer frame:
//      0xAA | ver(1) | declLen(2 LE) | field(2 LE) | crc16(2 LE, CRC16-MODBUS
//      of the preceding 5 bytes)
//      then declLen bytes of inner: type | seq | cmd | b3 | payload | crc32(4 LE)
//

import Foundation

public enum FrameError: Error, CustomStringConvertible {
    case tooShort(Int)
    case badMagic(UInt8)
    case declaredLengthExceedsBuffer(declared: Int, available: Int)
    case innerTooShort

    public var description: String {
        switch self {
        case .tooShort(let n): return "frame too short: \(n) bytes"
        case .badMagic(let b): return String(format: "bad magic byte: 0x%02x (expected 0xAA)", b)
        case .declaredLengthExceedsBuffer(let d, let a):
            return "declared inner length \(d) exceeds available bytes (\(a))"
        case .innerTooShort: return "inner payload shorter than type+seq+cmd+b3+crc32"
        }
    }
}

public enum Framing {
    public static let outerMagic: UInt8 = 0xAA

    /// CRC16-MODBUS: poly 0xA001 (reflected 0x8005), init 0xFFFF.
    public static func crc16Modbus<D: Sequence>(_ data: D) -> UInt16 where D.Element == UInt8 {
        var crc: UInt16 = 0xFFFF
        for b in data {
            crc ^= UInt16(b)
            for _ in 0..<8 {
                if crc & 1 != 0 {
                    crc = (crc >> 1) ^ 0xA001
                } else {
                    crc >>= 1
                }
            }
        }
        return crc
    }

    /// Standard CRC-32 (same polynomial zlib/gzip use), computed without
    /// linking zlib so this stays portable across all Apple platforms.
    private static let crc32Table: [UInt32] = {
        (0...255).map { i -> UInt32 in
            var c = UInt32(i)
            for _ in 0..<8 {
                c = (c & 1 != 0) ? (0xEDB88320 ^ (c >> 1)) : (c >> 1)
            }
            return c
        }
    }()

    public static func crc32Std<D: Sequence>(_ data: D) -> UInt32 where D.Element == UInt8 {
        var crc: UInt32 = 0xFFFFFFFF
        for b in data {
            let idx = Int((crc ^ UInt32(b)) & 0xFF)
            crc = crc32Table[idx] ^ (crc >> 8)
        }
        return crc ^ 0xFFFFFFFF
    }
}

public struct InnerMessage: Equatable {
    public let type: UInt8
    public let seq: UInt8
    public let cmd: UInt8
    public let b3: UInt8
    public let payload: Data
    public let crc32OK: Bool

    public init(type: UInt8, seq: UInt8, cmd: UInt8, b3: UInt8, payload: Data, crc32OK: Bool) {
        self.type = type; self.seq = seq; self.cmd = cmd; self.b3 = b3
        self.payload = payload; self.crc32OK = crc32OK
    }
}

public struct OuterFrame {
    public let version: UInt8
    public let field: UInt16
    public let headerCRCOK: Bool
    public let inner: InnerMessage
    public let raw: Data
}

/// Build an outbound 0xAA-framed command (matches puffinCommandFrame).
public func buildCommandFrame(msgType: UInt8, seq: UInt8, cmd: UInt8, b3: UInt8,
                               payload: Data = Data(), field: UInt16 = 0x0100,
                               version: UInt8 = 0x01) -> Data {
    var inner = Data([msgType, seq, cmd, b3])
    inner.append(payload)
    let innerCRC = Framing.crc32Std(inner)

    var headerBody = Data([version])
    headerBody.append(contentsOf: withUnsafeBytes(of: UInt16(inner.count + 4).littleEndian) { Data($0) })
    headerBody.append(contentsOf: withUnsafeBytes(of: field.littleEndian) { Data($0) })
    let headerCRC = Framing.crc16Modbus(headerBody)

    var out = Data([Framing.outerMagic])
    out.append(headerBody)
    out.append(contentsOf: withUnsafeBytes(of: headerCRC.littleEndian) { Data($0) })
    out.append(inner)
    out.append(contentsOf: withUnsafeBytes(of: innerCRC.littleEndian) { Data($0) })
    return out
}

/// Parse one complete 0xAA-framed message.
public func parseFrame(_ data: Data) throws -> OuterFrame {
    guard data.count >= 9 else { throw FrameError.tooShort(data.count) }
    let bytes = [UInt8](data)
    guard bytes[0] == Framing.outerMagic else { throw FrameError.badMagic(bytes[0]) }

    let version = bytes[1]
    let declLen = UInt16(bytes[2]) | (UInt16(bytes[3]) << 8)
    let field = UInt16(bytes[4]) | (UInt16(bytes[5]) << 8)
    let headerCRCRx = UInt16(bytes[6]) | (UInt16(bytes[7]) << 8)

    let headerBody = bytes[1..<6]
    let headerCRCOK = Framing.crc16Modbus(headerBody) == headerCRCRx

    let innerStart = 8
    let innerEnd = innerStart + Int(declLen)
    guard bytes.count >= innerEnd else {
        throw FrameError.declaredLengthExceedsBuffer(declared: Int(declLen),
                                                       available: bytes.count - innerStart)
    }

    let innerFull = Array(bytes[innerStart..<innerEnd])
    guard innerFull.count >= 4 + 4 else { throw FrameError.innerTooShort }

    let innerBody = innerFull[0..<(innerFull.count - 4)]
    let crcBytes = innerFull[(innerFull.count - 4)...]
    let crcRx = UInt32(crcBytes[crcBytes.startIndex]) |
        (UInt32(crcBytes[crcBytes.startIndex + 1]) << 8) |
        (UInt32(crcBytes[crcBytes.startIndex + 2]) << 16) |
        (UInt32(crcBytes[crcBytes.startIndex + 3]) << 24)
    let crc32OK = Framing.crc32Std(innerBody) == crcRx

    let msgType = innerBody[innerBody.startIndex]
    let seq = innerBody[innerBody.startIndex + 1]
    let cmd = innerBody[innerBody.startIndex + 2]
    let b3 = innerBody[innerBody.startIndex + 3]
    let payload = Data(innerBody[(innerBody.startIndex + 4)...])

    let inner = InnerMessage(type: msgType, seq: seq, cmd: cmd, b3: b3,
                              payload: payload, crc32OK: crc32OK)
    return OuterFrame(version: version, field: field, headerCRCOK: headerCRCOK,
                       inner: inner, raw: Data(bytes[0..<innerEnd]))
}

/// Streaming reassembler: feed it bytes as BLE notifications arrive (they
/// can be split across MTUs), get back any complete frames found so far.
/// Consumed bytes are removed from `buffer`; a trailing partial frame is
/// left in place for the next call.
public final class FrameReassembler {
    private var buffer = Data()

    public init() {}

    public func feed(_ chunk: Data) -> [OuterFrame] {
        buffer.append(chunk)
        var frames: [OuterFrame] = []

        while true {
            guard let magicIdx = buffer.firstIndex(of: Framing.outerMagic) else {
                buffer.removeAll()
                break
            }
            if magicIdx > buffer.startIndex {
                buffer.removeSubrange(buffer.startIndex..<magicIdx)
            }
            guard buffer.count >= 8 else { break }

            let b = [UInt8](buffer)
            let declLen = Int(UInt16(b[2]) | (UInt16(b[3]) << 8))
            let totalLen = 8 + declLen
            guard buffer.count >= totalLen else { break }

            let chunkData = buffer.prefix(totalLen)
            buffer.removeFirst(totalLen)

            do {
                frames.append(try parseFrame(Data(chunkData)))
            } catch {
                // Resync: drop the magic byte we just consumed and keep scanning.
                continue
            }
        }
        return frames
    }
}

File: Tools/whoop5_decoder/framing.py
"""
WHOOP 5.0 / MG ("puffin") BLE frame codec.

Ported from the publicly documented protocol used by the NOOP interoperability
project (github.com/NoopApp/noop) and its cited sources (b-nnett/goose,
Asherlc/dofek, judes.club's WHOOP5 writeup). This only implements the
*framing* layer -- getting bytes off the wire into (type, seq, cmd, b3,
payload) tuples with integrity verified. Field-level meaning of each payload
lives in messages.py, and only covers what's been publicly reverse engineered.

Outer frame:
    0xAA | ver(1) | declLen(2, LE) | field(2, LE, usually 0x0100) | crc16(2, LE)
    -- crc16 is CRC16-MODBUS over the preceding 6 header bytes --
    then declLen bytes of "inner" payload:
        type(1) | seq(1) | cmd(1) | b3(1) | payload(...) | crc32(4, LE)
    -- crc32 is CRC32 (zlib/standard) over [type..payload], i.e. inner minus
       its own trailing 4 crc bytes --

This matches what NOOP calls Framing.puffinCommandFrame.
"""

from __future__ import annotations

import struct
import zlib
from dataclasses import dataclass
from typing import Optional

OUTER_MAGIC = 0xAA
HEADER_LEN = 7          # magic, ver, declLen(2), field(2), crc16(2) minus magic... see below
INNER_MIN_LEN = 4        # type, seq, cmd, b3
CRC_LEN = 4               # trailing crc32


class FrameError(ValueError):
    pass


def crc16_modbus(data: bytes) -> int:
    """CRC16-MODBUS: poly 0xA001 (reflected 0x8005), init 0xFFFF."""
    crc = 0xFFFF
    for b in data:
        crc ^= b
        for _ in range(8):
            if crc & 1:
                crc = (crc >> 1) ^ 0xA001
            else:
                crc >>= 1
    return crc & 0xFFFF


def crc32_std(data: bytes) -> int:
    """Standard zlib CRC-32 over the inner message (type..payload)."""
    return zlib.crc32(data) & 0xFFFFFFFF


@dataclass
class InnerMessage:
    type: int          # e.g. 0x23 for a puffin command/response envelope
    seq: int
    cmd: int
    b3: int             # 0x01 for GET_HELLO/SET_CONFIG, 0x00 for GET_DATA_RANGE/SEND_HISTORICAL
    payload: bytes
    crc32_ok: bool

    def raw_inner(self) -> bytes:
        return bytes([self.type, self.seq, self.cmd, self.b3]) + self.payload


@dataclass
class OuterFrame:
    version: int
    field: int
    header_crc_ok: bool
    inner: InnerMessage
    raw: bytes


def build_command_frame(msg_type: int, seq: int, cmd: int, b3: int,
                         payload: bytes = b"", field: int = 0x0100,
                         version: int = 0x01) -> bytes:
    """Build an outbound 0xAA-framed command, matching puffinCommandFrame."""
    inner = bytes([msg_type, seq, cmd, b3]) + payload
    inner_crc = struct.pack("<I", crc32_std(inner))
    inner_full = inner + inner_crc

    decl_len = len(inner_full)
    header_body = struct.pack("<BH H", version, decl_len, field) if False else (
        bytes([version]) + struct.pack("<H", decl_len) + struct.pack("<H", field)
    )
    header_crc = struct.pack("<H", crc16_modbus(header_body))

    return bytes([OUTER_MAGIC]) + header_body + header_crc + inner_full


def parse_frame(data: bytes) -> OuterFrame:
    """Parse one complete 0xAA-framed message (outer header + inner + crc32)."""
    if len(data) < 9:
        raise FrameError(f"frame too short: {len(data)} bytes")
    if data[0] != OUTER_MAGIC:
        raise FrameError(f"bad magic byte: 0x{data[0]:02x} (expected 0xAA)")

    version = data[1]
    decl_len = struct.unpack_from("<H", data, 2)[0]
    field = struct.unpack_from("<H", data, 4)[0]
    header_crc_rx = struct.unpack_from("<H", data, 6)[0]

    header_body = data[1:6]
    header_crc_ok = crc16_modbus(header_body) == header_crc_rx

    inner_start = 8
    inner_end = inner_start + decl_len
    if len(data) < inner_end:
        raise FrameError(
            f"declared inner length {decl_len} exceeds available bytes "
            f"({len(data) - inner_start})"
        )

    inner_full = data[inner_start:inner_end]
    if len(inner_full) < INNER_MIN_LEN + CRC_LEN:
        raise FrameError("inner payload shorter than type+seq+cmd+b3+crc32")

    inner_body = inner_full[:-CRC_LEN]
    crc_rx = struct.unpack_from("<I", inner_full, len(inner_full) - CRC_LEN)[0]
    crc32_ok = crc32_std(inner_body) == crc_rx

    msg_type, seq, cmd, b3 = inner_body[0], inner_body[1], inner_body[2], inner_body[3]
    payload = inner_body[4:]

    inner = InnerMessage(type=msg_type, seq=seq, cmd=cmd, b3=b3,
                          payload=payload, crc32_ok=crc32_ok)
    return OuterFrame(version=version, field=field, header_crc_ok=header_crc_ok,
                       inner=inner, raw=data[:inner_end])


def iter_frames(buf: bytearray):
    """
    Generator that pulls as many complete frames as possible out of a
    streaming byte buffer (BLE notifications can arrive split across
    multiple packets/MTUs), yielding OuterFrame and trimming consumed
    bytes from `buf` in place. Leaves a trailing partial frame in `buf`.
    """
    while True:
        # find magic byte
        try:
            start = buf.index(OUTER_MAGIC)
        except ValueError:
            buf.clear()
            return
        if start > 0:
            del buf[:start]

        if len(buf) < 8:
            return  # not enough for header yet

        decl_len = struct.unpack_from("<H", buf, 2)[0]
        total_len = 8 + decl_len
        if len(buf) < total_len:
            return  # wait for more bytes

        chunk = bytes(buf[:total_len])
        del buf[:total_len]
        try:
            yield parse_frame(chunk)
        except FrameError:
            # resync: drop the magic byte we just consumed and keep scanning
            continue
